backend/src/main/kotlin/app/ehrenamtskarte/backend/verification/Argon2IdHasher.kt

```diff
@@ -0,0 +1,80 @@
+import app.ehrenamtskarte.backend.common.utils.Environment
+import app.ehrenamtskarte.backend.common.webservice.KOBLENZ_PEPPER_SYS_ENV
+import app.ehrenamtskarte.backend.verification.CanonicalJson
+import org.apache.commons.codec.binary.Hex
+import org.bouncycastle.crypto.generators.Argon2BytesGenerator
+import org.bouncycastle.crypto.params.Argon2Parameters
+import java.nio.charset.StandardCharsets
+import java.util.Base64
+
+class Argon2IdHasher {
+    companion object {
+        /**
+         * Copied from spring-security Argon2EncodingUtils.java licenced under Apache 2.0
+         *
+         * Encodes a raw Argon2-hash and its parameters into the standard Argon2-hash-string
+         * as specified in the reference implementation
+         * (https://github.com/P-H-C/phc-winner-argon2/blob/master/src/encoding.c#L244):
+         *
+         * {@code $argon2<T>[$v=<num>]$m=<num>,t=<num>,p=<num>$<bin>$<bin>}
+         **/
+        @Throws(IllegalArgumentException::class)
+        fun encode(hash: ByteArray?, parameters: Argon2Parameters): String? {
+            val b64encoder: Base64.Encoder = Base64.getEncoder().withoutPadding()
+            val stringBuilder = StringBuilder()
+            val type = when (parameters.type) {
+                Argon2Parameters.ARGON2_d -> "\$argon2d"
+                Argon2Parameters.ARGON2_i -> "\$argon2i"
+                Argon2Parameters.ARGON2_id -> "\$argon2id"
+                else -> throw IllegalArgumentException("Invalid algorithm type: " + parameters.type)
+            }
+            stringBuilder.append(type)
+            stringBuilder.append("\$v=")
+                .append(parameters.version)
+                .append("\$m=")
+                .append(parameters.memory)
+                .append(",t=")
+                .append(parameters.iterations)
+                .append(",p=")
+                .append(parameters.lanes)
+            if (parameters.salt != null) {
+                stringBuilder.append("$").append(b64encoder.encodeToString(parameters.salt))
+            }
+            stringBuilder.append("$").append(b64encoder.encodeToString(hash))
+            return stringBuilder.toString()
+        }
+
+        fun hashUserData(cardInfo: Card.CardInfo): String? {
+            val canonicalJson = CanonicalJson.messageToMap(cardInfo)
+            val hashLength = 32
+            if (!isCanonicalJsonValid(canonicalJson)) {
+                throw Exception("Invalid Json input for hashing")
+            }
+
+            val pepper = Environment.getVariable(KOBLENZ_PEPPER_SYS_ENV) //TODO handle if Null
```

Comment:
What error handling do we want, if the env var cannot be accesed / is not set?

Reply:
just added a basic exception. Do we need a separate exception for that?